### PR TITLE
Add module for adding/deleting computers via MS-SAMR

### PR DESCRIPTION
**Pull request by zeroSteiner**
This adds a new auxiliary/admin module that can be used to add, lookup and delete computer accounts from an active directory domain. By default standard AD users can add up to 10 computers to the domain. When a user has access to do this (such as via plaintext credentials at the moment), the computer account can offer a sort of foothold into the domain. It's also becoming a common attack primitive in [certain scenarios](https://cravaterouge.github.io/ad/privesc/2022/05/11/bloodyad-and-CVE-2022-26923.html).

Because there's typically a limited number of computers a user can add, the module's `ADD_COMPUTER` action will fail with `STATUS_DS_MACHINE_ACCOUNT_QUOTA_EXCEEDED` if used repeatedly. Also, while a standard user can create a computer account, additional privileges are required to delete it.

This requires the changes from rapid7/ruby_smb#231

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use auxiliary/admin/dcerpc/samr_computer`
- [ ] Set the `RHOSTS`, `SMBUser` and `SMBPass` options (`SMBDomain` is only required if the target has more than one domain on it)
- [ ] Run the `ADD_COMPUTER` action
    - [ ] A new computer will be created in the domain, the password printed to the screen and stored in the credentials table (as visible with `creds`)
- [ ] Set `COMPUTER_NAME` to the name of the new computer account
- [ ] Run the `LOOKUP_COMPUTER` action, see that the computer exists, and the SID is resolved
- [ ] Run the `DELETE_COMPUTER` action, see that the computer is deleted (**this action requires more privileges**, use a DA or other elevated account)
- [ ] Run the `LOOKUP_COMPUTER` action, see that the computer does not exist any more

The `ADD_COMPUTER` action makes a DCERPC request to set the password that uses application-layer encryption using either the session key or the application key from the underlying SMB transport. I tested this action using SMB 2, and SMB 3 both with and without transport-level encryption. The application key is only set for SMBv3 dialects. The module will use the application key when it's generated otherwise it'll use the session key. The SMB version and encryption settings can be configured through the `SMB::ProtocolVersion` and `SMB::AlwaysEncrypt` advanced datastore options.

## Demo

```
msf6 auxiliary(admin/dcerpc/samr_computer) > set RHOSTS 192.168.159.96
RHOSTS => 192.168.159.96
msf6 auxiliary(admin/dcerpc/samr_computer) > set SMBUser aliddle
SMBUser => aliddle
msf6 auxiliary(admin/dcerpc/samr_computer) > set SMBPass Password1
SMBPass => Password1
msf6 auxiliary(admin/dcerpc/samr_computer) > show options 
Module options (auxiliary/admin/dcerpc/samr_computer):
   Name           Current Setting  Required  Description
   ----           ---------------  --------  -----------
   COMPUTER_NAME                   no        The computer name
   RHOSTS         192.168.159.96   yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
   RPORT          445              yes       The target port (TCP)
   SMBDomain      .                no        The Windows domain to use for authentication
   SMBPass        Password1        no        The password for the specified username
   SMBUser        aliddle          no        The username to authenticate as
Auxiliary action:
   Name          Description
   ----          -----------
   ADD_COMPUTER  Add a computer account
msf6 auxiliary(admin/dcerpc/samr_computer) > run
[*] Running module against 192.168.159.96
[*] 192.168.159.96:445 - Using automatically identified domain: MSFLAB
[+] 192.168.159.96:445 - Successfully created MSFLAB\DESKTOP-2X8F54QG$ with password MCoDkNALd3SdGR1GoLhqniEkWa8Me9FY
[*] Auxiliary module execution completed
msf6 auxiliary(admin/dcerpc/samr_computer) > creds
Credentials
===========
host            origin          service        public             private                           realm   private_type  JtR Format
----            ------          -------        ------             -------                           -----   ------------  ----------
192.168.159.96  192.168.159.96  445/tcp (smb)  DESKTOP-2X8F54QG$  MCoDkNALd3SdGR1GoLhqniEkWa8Me9FY  MSFLAB  Password      
msf6 auxiliary(admin/dcerpc/samr_computer) >
```

### CODE REVIEW

**Review by cdelafuente-r7 — COMMENTED** [review]
Thanks @zeroSteiner for this module! I did a first round of tests against a Windows Server 2019 DC and it works great. I will continue testing against other targets/protocol versions. In the meantime, I left a few comments for you to review when you get a chance.

File: documentation/modules/auxiliary/admin/dcerpc/samr_computer.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+Add, lookup and delete computer accounts via MS-SAMR. By default standard active directory users can add up to 10 new
+computers to the domain. Administrative privileges however are required to delete the created accounts.
+
+## Verification Steps
+
+1. From msfconsole
+2. Do: `use auxiliary/admin/dcerpc/samr_computer`
+3. Set the `RHOSTS`, `SMBUser` and `SMBPass` options
+   1. Set the `COMPUTER_NAME` option for `DELETE_COMPUTER` and `LOOKUP_COMPUTER` actions
+4. Run the module and see that a new machine account was added
```
[the remark this answers is not in the document]
Comment (zeroSteiner) [reply documentation/modules/auxiliary/admin/dcerpc/samr_computer.md:11]: I wanted that to be a substep, so all the set-option instructions are together but `COMPUTER_NAME` is only required for those actions.

Verified it renders as intended:
image

File: modules/auxiliary/admin/dcerpc/samr_computer.rb [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+      tag: RubySMB::Dcerpc::Samr::USER_INTERNAL4_INFORMATION_NEW,
+      member: RubySMB::Dcerpc::Samr::SamprUserInternal4InformationNew.new(
+        i1: {
+          password_expired: 1,
+          which_fields: RubySMB::Dcerpc::Samr::USER_ALL_NTPASSWORDPRESENT | RubySMB::Dcerpc::Samr::USER_ALL_PASSWORDEXPIRED
+        },
+        user_password: {
+          buffer: RubySMB::Dcerpc::Samr::SamprEncryptedUserPasswordNew.encrypt_password(
+            password,
+            @simple.client.application_key.blank? ? @simple.client.session_key : @simple.client.application_key
```
[the remark this answers is not in the document]
Comment (zeroSteiner) [reply modules/auxiliary/admin/dcerpc/samr_computer.rb]: Fixed with commits https://github.com/rapid7/metasploit-framework/pull/16677/commits/41ba2d263b06d1d45d6a642a69030f576a5ce56d and https://github.com/rapid7/ruby_smb/pull/231/commits/9fafecabacce58d71fd6028dbd073fde5d345db9.

The application key is now always set to a usable value after authenticating by defaulting it to the session key, only updating it when the dialect is v3.

**Review by cdelafuente-r7 — APPROVED** [review]

**Comment by cdelafuente-r7** [comment]
# Release Notes
This adds an auxiliary module that can be used to add, lookup and delete computer accounts from an active directory domain. The computer account can offer a sort of foothold into the domain for lateral movements or as a common attack primitive.

**Pull request merged**